rest/indextest/index_test.go

```diff
@@ -596,6 +596,226 @@ func TestSyncOnline(t *testing.T) {
+	sc, closeFn := rest.StartServerWithConfig(t, &bootstrapConfig)
+	defer closeFn()
+
+	// Set testing callbacks for async initialization
+	collectionCount := int64(0)
+	initStarted := make(chan error)
+	unblockInit := make(chan error)
+	collectionCompleteCallback := func(dbName, collectionName string) {
+		count := atomic.AddInt64(&collectionCount, 1)
+		// On first collection, close initStarted channel
+		log.Printf("collection callback count: %v", count)
+		if count == 1 {
+			log.Printf("closing initStarted")
+			close(initStarted)
+		}
+		rest.WaitForChannel(t, unblockInit, "waiting for test to unblock initialization")
+	}
+	sc.DatabaseInitManager.SetCallbacks(collectionCompleteCallback, nil)
+
+	ctx := base.TestCtx(t)
+	// Get a test bucket, and use it to create the database.
+	tb := base.GetTestBucket(t)
+	defer tb.Close(ctx)
+
+	importFilter := "function(doc) { return true }"
+	syncFunc := "function(doc){ channel(doc.channels); }"
+
+	dbName := "db"
+	dbConfig := makeDbConfig(t, tb, syncFunc, importFilter)
+	dbConfig.Name = dbName
+	dbConfig.StartOffline = base.BoolPtr(true)
+
+	keyspace := dbName
+	if len(dbConfig.Scopes) > 0 {
+		keyspaces := getRESTKeyspaces(dbName, dbConfig.Scopes)
+		keyspace = keyspaces[0]
+	}
+
+	bucketName := tb.GetName()
+	groupID := sc.Config.Bootstrap.ConfigGroupID
+
+	// Simulate creation of the database with offline=true on another node
+	version, err := rest.GenerateDatabaseConfigVersionID(ctx, "", &dbConfig)
+	require.NoError(t, err)
+	metadataID, err := sc.BootstrapContext.ComputeMetadataIDForDbConfig(ctx, &dbConfig)
+	require.NoError(t, err)
+
+	databaseConfig := dbConfig.ToDatabaseConfig()
+	databaseConfig.Version = version
+	databaseConfig.MetadataID = metadataID
+
+	_, err = sc.BootstrapContext.InsertConfig(ctx, bucketName, groupID, databaseConfig)
+	require.NoError(t, err)
+
+	// Wait for init to start before interacting with the db, validate db state is offline
+	rest.WaitForChannel(t, initStarted, "waiting for initialization to start")
+	log.Printf("initialization started")
+	waitAndRequireDBState(t, sc, dbName, db.DBOffline)
+
+	log.Printf("keyspace: %v", keyspace)
+
+	// Update the bucket config to bring the database online
+	_, err = sc.BootstrapContext.UpdateConfig(ctx, bucketName, groupID, dbName, func(bucketDbConfig *rest.DatabaseConfig) (updatedConfig *rest.DatabaseConfig, err error) {
+		bucketDbConfig.StartOffline = base.BoolPtr(false)
+		return bucketDbConfig, nil
+	})
+	require.NoError(t, err)
+	waitAndRequireDBState(t, sc, dbName, db.DBStarting)
+
+	// Attempt to update import filter while in starting mode
+	importFilter = "function(doc){ return false; }"
+	resp := rest.BootstrapAdminRequest(t, sc, http.MethodPut, "/"+keyspace+"/_config/import_filter", importFilter)
+	resp.RequireStatus(http.StatusOK)
+
+	resp = rest.BootstrapAdminRequest(t, sc, http.MethodGet, "/"+keyspace+"/_config/import_filter", "")
+	resp.RequireResponse(http.StatusOK, importFilter)
```

Comment:
I'd prefer to make sure we don't miss steps here, and put the test cases for sync and import_filter and make sure Put/Get/Delete/Get workflow works. This could be pulled out from the code above.

Reply:
What steps are you thinking we're missing here?  This test (TestAsyncInitRemoteConfigUpdates) is ensuring that async works correctly when the db is updated on a remote nod, while the previous test (TestAsyncInitConfigUpdates) validates all the different config update options locally.  I included the import filter change here to make sure the database wasn't in a state that would reject changes after processing a remote update.  Beyond that, I'd expect the coverage in TestAsyncInitRemoteConfigUpdates to suffice.

Reply:
I think we can drop this, but I thought we'd just be complete to keep import and sync. I do think that the combination of tests covers all cases.